recipes/dns_config.rb

```diff
@@ -0,0 +1,148 @@
+  else
+    execute "hostnamectl set-hostname #{hostname}"
+  end
+end
+
+# Function to extend DNS resolver configuration by appending the Route53 domain created within the cluster
+# ($CLUSTER_NAME.pcluster) and be listed as a "search" domain in the resolv.conf file.
+def append_cluster_dns_domain
+  if platform?('ubuntu') && node['platform_version'] == "18.04"
+    Chef::Log.info("Appending search domain '#{node['cfncluster']['cfn_dns_domain']}' to /etc/systemd/resolved.conf")
+    # Configure resolved to automatically append Route53 search domain in resolv.conf.
+    # On Ubuntu18 resolv.conf is managed by systemd-resolved.
+    # TODO verify if Ubuntu Desktop uses NetworkManager and we have to do a custom configuration for custom amis using this OS
+    replace_or_add "Append Route53 search domain in /etc/systemd/resolved.conf" do
+      path "/etc/systemd/resolved.conf"
+      pattern "Domains=*"
+      line "Domains=#{node['cfncluster']['cfn_dns_domain']}"
+    end
+  else
+    Chef::Log.info("Appending search domain '#{node['cfncluster']['cfn_dns_domain']}' to /etc/dhcp/dhclient.conf")
+    # Configure dhclient to automatically append Route53 search domain in resolv.conf
+    # - on CentOS6 resolv.conf is managed by network + dhclient,
+    # - on CentOS7, Alinux and Alinux2 by NetworkManager + dhclient,
+    # - on Ubuntu16 by networking + dhclient
+    replace_or_add "Append Route53 search domain in /etc/dhcp/dhclient.conf" do
+      path "/etc/dhcp/dhclient.conf"
+      pattern "append domain-name*"
+      line "append domain-name \" #{node['cfncluster']['cfn_dns_domain']}\";"
+    end
+
+    if platform?('centos') && node['platform_version'].to_i < 7
+      # On CentOS6 there is a dhclient configuration file for eth0
+      replace_or_add "Append Route53 search domain in /etc/dhcp/dhclient-eth0.conf" do
+        path "/etc/dhcp/dhclient-eth0.conf"
+        pattern "append domain-name*"
+        line "append domain-name \" #{node['cfncluster']['cfn_dns_domain']}\";"
+      end
+    end
+  end
+end
+
+# Utility method to restart network service according to the OS.
+def restart_network_service
+  network_service_name = value_for_platform(
+    ['centos'] => {
+      '<7.0' => 'network'
+    },
+    ['amazon'] => {
+      '>=2013' => 'network' # Alinux1
+    },
+    %w[ubuntu debian] => {
+      '16.04' => 'networking',
+      '>=18.04' => 'systemd-resolved'
+    },
+    'default' => 'NetworkManager'
+  )
+  Chef::Log.info("Restarting '#{network_service_name}' service, for platform #{node['platform']} '#{node['platform_version']}'")
+  service network_service_name.to_s do
+    action %i[restart]
+  end
+end
+
+if node['cfncluster']['cfn_scheduler'] == 'slurm'
+  # Heterogeneous Instance Type
+  if node['cfncluster']['cfn_dns_domain'].nil? ||
+     node['cfncluster']['cfn_dns_domain'].empty? ||
+     node['cfncluster']['cfn_dns_domain'] == "NONE"
+    Chef::Log.info("DNS domain not configured")
+    return
```

Comment:
I think we can't just return. We should consider the following:
- option 1: the user disables DNS resolution (which will more likely be a configuration parameter). In this case we still configure the hostname as the Slurm nodename but DNS resolution won't work. Users with custom DNS might want to use this option
- option 2: the user disables DNS resolution but also wants to preserve the old way of configuring hostnames (this can be set with an extra_json parameter). In this case we need to still configure hostname the old fashion.

Reply:
Yes, I'm going to add and test the two configuration parameters